src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/XmlQueryContext.cs

```diff
@@ -230,6 +232,7 @@ public object GetLateBoundObject(string namespaceUri)
         /// <summary>
         /// Return true if the late bound object identified by "namespaceUri" contains a method that matches "name".
         /// </summary>
+        [RequiresUnreferencedCode(Scripts.ExtensionFunctionCannotBeStaticallyAnalyzed)]
         public bool LateBoundFunctionExists(string name, string namespaceUri)
```

Comment:
My understanding is:
* The only way to get any late bound stuff to happen is to call `XsltArgumentList.AddExtensionObject` (my reading of the code says that as well).
* `XsltArgument.AddExtensionObject` is marked with RUC (and will probably remain that way until we can annotate instance types, which we currently can't).
* If given app doesn't use late bound stuff - it will never call `AddExtensionObject` -> no warning there - and thus we want to avoid emitting warning in the rest of the codebase for this reason.
* If given app does use late bound stuff - it MUST call `AddExtensionObject` and will get a warning there.

In other places we suppress warnings based on the above logic (FunctionQuery is like that), but here we propagate it. I understand that this is a public API, but on its own it's not dangerous - the implementation which allows extension objects is dangerous.

I know it's a subjective view, but I think it would good to be consistent - if we internally rely on AddExtensionObject being the "bad" entrypoint, then let's stick to that everywhere:
* I would mark `XsltArgumentList.GetExtensionObejct` with RUC (so that it clearly points out the places where we access the dangerous stuff) and suppress the callsites.

Alternatively - we could turn this into a feature switch - put the AddExtensionObject behind a feature switch (that is, is the switch the turned off, this method always throws - we don't allow extension objects).
* I would maybe even do this, regardless if we expose the feature switch or not
* It would make us draw clear line where the late bound stuff happens and cleanly cut it away
* Applying annotations on this would then be easier and cleaner

Reply:
> Alternatively - we could turn this into a feature switch - put the AddExtensionObject behind a feature switch (that is, is the switch the turned off, this method always throws - we don't allow extension objects).

I'm not sure using a feature switch here is useful. Typically we use feature switches in places where we don't have a public API telling us the developer's intent. Here if they call `AddExtensionObject`, the intent is clear - they are trying to do something that is not trim compatible - so let's just warn as usual.

I agree with the rest of the points - we should be consistent on suppressing the warnings if they can only be reached when someone calls `AddExtensionObject`.

Reply:
I probably wasn't clear .. I was thinking something like "Think what we would do if we were to add a feature switch for this, and design the annotations or refactor the code in that way". The idea is that it would introduce a nicer "cut line". I agree that we should not make it a real feature switch.

Reply:
@vitek-karas This is more or less the way I went attacking this specific warning:
- Started with the callsite that had the linker warning, which in this case, was XsltCompileContext.GetExtensionMethod which was trying to call GetMethods() on the extension object passed in through the arglist.
- Worked my way up the stack and annotated as RUC all of the part of the stack that was specific to parsing the extension function call.
- Once I reached the point where the call stack wasn't specifically about extension methods, and instead just about transforming the next node of the stylesheet (which happened to be an extension method) then I suppressed the message saying that AddExtensionObject was already annotated. The reasoning behind was that I knew that all of the part that involved ExtensionMethods will always be unsafe, whereas the rest of the call stack would only be unsafe if AddExtensionObject was called.

All that said, I think I agree with your approach and would also be cleaner if we just annotated GetExtensionObject and then suppressed the callsites with message saying that AddExtensionObject is already annotated. I'll make these changes and push them to this PR. Also, let me know if I misunderstood something here.

Reply:
I agree that as a way to get there, your approach is good - it unfortunately makes the code less maintainable (in my opinion) because it creates "Dependencies" between very different parts of the code base (having the ability to introduce new functions to the transform is in itself not dangerous - but we have to point to a completely different part of the codebase to say - well this specific implementation uses reflection and it's safe because of yet another thing over there).

This is one of the reasons linker makes it tough to put RUC on a virtual method (it then forces it on all implementations). The fact that you annotated a perfectly fine method (which just throws) as RUC should be a warning in itself. Unfortunately the tooling is there around this (and I don't know if it's even possible).